### PR TITLE
Fix wrong min-max filter result of parquet reader

### LINKED ISSUE #4390: result error when query from HUDI external table
### Steps to reproduce the behavior (Required)
1. create HUDI table using parquet file format and gzip compression，and with 1G data inserted.
2. create HUDI external table;
3. run tpch and tpcds querys, and some querys' result are not consistency with OLAP tables;

### Expected behavior (Required)
tpch q14.sql:
```
mysql> select   100.00 * sum(case     when p_type like 'PROMO%'       then l_extendedprice * (1 - l_discount)     else 0   end) / sum(l_extendedprice * (1 - l_discount)) as promo_revenue from   lineitem,   part where   l_partkey = p_partkey   and l_shipdate >= date '1995-09-01'   and l_shipdate < date '1995-09-01' + interval '1' month;
+---------------+
| promo_revenue |
+---------------+
| 16.3807786264 |
+---------------+
1 row in set (0.11 sec)
```

### Real behavior (Required)
```
mysql> select   100.00 * sum(case     when p_type like 'PROMO%'       then l_extendedprice * (1 - l_discount)     else 0   end) / sum(l_extendedprice * (1 - l_discount)) as promo_revenue from   lineitem,   part where   l_partkey = p_partkey   and l_shipdate >= date '1995-09-01'   and l_shipdate < date '1995-09-01' + interval '1' month;
+---------------+
| promo_revenue |
+---------------+
|          NULL |
+---------------+
1 row in set (0.08 sec)
```

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
```
mysql> select current_version();
+------------------------+
| current_version()      |
+------------------------+
| QA_TEST_MASTER ac2c40e |
+------------------------+
1 row in set (0.00 sec)
```

### LINKED ISSUE #4555: result error from HUDI external table when date type as where condition
### Steps to reproduce the behavior (Required)
1. create hudi table and insert data:
```
 create table hudi_parquet_gzip(
 uuid int,
 col_boolean boolean,
 col_int int,
 col_long long,
 col_float float,
 col_double double,
 col_decimal decimal(38,18),
 col_date date,
 col_timestamp timestamp,
 col_string string,
 col_binary binary,
 col_array array<int>,
 col_struct struct<a:string,b:int>,
 col_map map<string,int>
 ) USING hudi 
 TBLPROPERTIES(
 type = 'cow',
 primarykey = 'uuid');

 insert into hudi_parquet_gzip select 1,true,1,1.01,1.001,1.0001,10000000000.0000000001,cast('2020-01-01' as date),cast('2020-01-01 00:00:01' as timestamp),'Top 10 Unsolved Mysteries of Paleontological Dinosaurs, Did You Know?',cast('1110001010101011001001' as binary),array(1,10,100),null,null;
 insert into hudi_parquet_gzip select 2,false,2,2.02,2.002,2.0002,20000000000.0000000002,cast('2020-02-01' as date),cast('2020-02-01 00:00:02' as timestamp),'Xi Jinping, general secretary of the Communist Party of China Central Committee, called on Tuesday for unrelenting efforts in exercising full and rigorous governance over the Party, saying that the CPC will continue to show zero tolerance for corruption.',cast('1110001010101011001001' as binary),array(2,20,200),null,null;
 insert into hudi_parquet_gzip select 3,true,3,3.03,3.003,3.0003,30000000000.0000000003,cast('2020-03-01' as date),cast('2020-03-01 00:00:04' as timestamp),'Nation raises caution on overseas packages',cast('1110001010101011001001' as binary),array(3,30,300),null,null;
 insert into hudi_parquet_gzip select 4,false,4,4.04,4.004,4.0004,40000000000.0000000004,cast('2020-04-01' as date),cast('2020-04-01 00:00:04' as timestamp),'Shanghai, Shenzhen register most newly listed firms in 2021',cast('1110001010101011001001' as binary),array(4,40,400),null,null;
 insert into hudi_parquet_gzip select 5,true,5,5.05,5.005,5.0005,50000000000.0000000005,cast('2020-05-01' as date),cast('2020-05-01 00:00:05' as timestamp),'A total of 4,685 companies have been listed on the A-share market in China as of Dec 31, 2021, with 46 percent of them based in Beijing, Shanghai, Shenzhen, Hangzhou, Suzhou, Guangzhou, Ningbo, Nanjing, Wuxi, and Chengdu, said a report of National Business Daily on Wednesday.',cast('1110001010101011001001' as binary),array(5,50,500),null,null;
 insert into hudi_parquet_gzip select 6,false,6,6.06,6.006,6.0006,60000000000.0000000006,cast('2020-06-01' as date),cast('2020-06-01 00:00:06' as timestamp),'The development of enterprises, as well as the push of local governments, has boosted the listings, the report said. For example, in early 2018, Central China''s Hubei province launched a plan to double its listed companies by including the listing of enterprises into the government''s annual performance assessment.',cast('1110001010101011001001' as binary),array(6,60,600),null,null;
 insert into hudi_parquet_gzip select 7,true,7,7.07,7.007,7.0007,70000000000.0000000007,cast('2020-07-01' as date),cast('2020-07-01 00:00:07' as timestamp),'Most listed companies in the top 10 cities come from industries such as information technology, electronics, mechanical equipment, medical biology, and electric power equipment, with information technology bearing most listed companies.',cast('1110001010101011001001' as binary),array(7,70,700),null,null;
 insert into hudi_parquet_gzip select 8,false,8,8.08,8.008,8.0008,80000000000.0000000008,cast('2020-08-01' as date),cast('2020-08-01 00:00:08' as timestamp),'Citi report finds MNC mood in China buoyant',cast('1110001010101011001001' as binary),array(8,80,800),null,null;
 insert into hudi_parquet_gzip select 9,true,9,9.09,9.009,9.0009,90000000000.0000000009,cast('2020-09-01' as date),cast('2020-09-01 00:00:09' as timestamp),'Xi: China, Russia major champions of multilateralism, global justice',cast('1110001010101011001001' as binary),array(9,90,900),null,null;
 insert into hudi_parquet_gzip select 10,false,10,10.10,10.010,10.00010,100000000000.00000000010,cast('2020-10-01' as date),cast('2020-10-01 00:00:10' as timestamp),'Volkswagen deliveries hit by chip shortages',cast('1110001010101011001001' as binary),array(10,100,1000),null,null;
 insert into hudi_parquet_gzip select 11,true,11,11.11,11.011,11.00011,110000000000.00000000011,cast('2020-11-01' as date),cast('2020-11-01 00:00:11' as timestamp),'Light installations featuring the tiger, the Chinese zodiac animal for the upcoming lunar year, will be set up in the main plaza.',cast('1110001010101011001001' as binary),array(11,110,1100),null,null;
 insert into hudi_parquet_gzip select 12,false,12,12.12,12.012,12.00012,120000000000.00000000012,cast('2020-12-01' as date),cast('2020-10-01 00:00:10' as timestamp),'The annual lantern show at Yuyuan Garden, a historic tourist site in Shanghai, will kick off on Tuesday and run through Feb 28.',cast('1110001010101011001001' as binary),array(12,120,1200),null,null;
 insert into hudi_parquet_gzip select 13,null,null,null,null,null,null,null,null,null,null,null,null,null;
 insert into hudi_parquet_gzip select 14,null,null,null,null,null,12345678901234567890.123456789012345678,null,null,'',null,null,null,null;
```

2. create hudi external table:
```
create external table ex_hudi_tbl8 (
 uuid int
 ,col_boolean boolean
 ,col_int int
 ,col_long bigint
 ,col_float float
 ,col_double double
 ,col_decimal decimal(38,18)
 ,col_date date
 ,col_string string
 ,col_binary varchar(200))
 ENGINE=hudi
 properties (
 "resource" = "hudi_emr_tn",
 "table" = "hudi_parquet_gzip",
 "database" = "hudi_db");
```

3. select col_date from hudi external table:
```
mysql> select col_date from ex_hudi_tbl8;
+------------+
| col_date   |
+------------+
| 2020-01-01 |
| 2020-02-01 |
| 2020-03-01 |
| 2020-04-01 |
| 2020-05-01 |
| 2020-06-01 |
| 2020-07-01 |
| 2020-08-01 |
| 2020-09-01 |
| 2020-10-01 |
| 2020-11-01 |
| 2020-12-01 |
| NULL       |
| NULL       |
+------------+
14 rows in set (0.03 sec)

mysql> select col_date from ex_hudi_tbl8 where col_date > '2020-10-01';
Empty set (0.02 sec)
```

### Expected behavior (Required)
results from hudi external table same with hudi table:
```
spark-sql> select col_date from hudi_parquet_gzip where col_date > '2020-10-01';
2020-11-01
2020-12-01
Time taken: 0.688 seconds, Fetched 2 row(s)
```

### Real behavior (Required)
```
mysql> select col_date from ex_hudi_tbl8 where col_date > '2020-10-01';
Empty set (0.02 sec)
```

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
```
mysql> select current_version();
+------------------------+
| current_version()      |
+------------------------+
| QA_TEST_MASTER 61b1f1f |
+------------------------+
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by miomiocat**
co-woker: @DorianZheng 

## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：
Closes #4555 
Closes #4390 

## Problem Summary(Required) ：

Currently proper convert and transformation is missing for min-max value of complex type like date or datetime in parquet reader and it will cause wrong min-max filter to use to skip parquet's row group and finally return the totally incorrect query results.

This PR extract converter from column reader and reuse it in min-max reader.

### CODE REVIEW

File: be/src/exec/parquet/column_reader.h [hunk 1]
```diff
@@ -27,6 +30,269 @@ struct ColumnReaderOptions {
     std::string timezone;
 };
 
+// When doing decimal convert, source scale may not equal with destination scale,
+// when destination scale is greater than source, source unscaled data will be scaled up
+// to match destination scale.
+enum DecimalScaleType {
```
Comment (DorianZheng) [inline be/src/exec/parquet/column_reader.h:36]: use `enum class` instead

File: be/src/exec/parquet/column_reader.h [hunk 2]
```diff
@@ -27,6 +30,269 @@ struct ColumnReaderOptions {
...
+    NO_SCALE,
```
Comment (DorianZheng) [inline be/src/exec/parquet/column_reader.h:37]: should start with `k`, such as `kNoScale`

File: be/src/exec/parquet/column_reader.h [hunk 3]
```diff
@@ -27,6 +30,269 @@ struct ColumnReaderOptions {
...
+    }
+
+private:
+    DecimalScaleType _scale_type = NO_SCALE;
+    DestPrimitiveType _scale_factor = 1;
+};
+
+class ColumnConverterFactory {
+public:
+    static Status init_converter(const ParquetField* field, const TypeDescriptor& typeDescriptor,
```
Comment (DorianZheng) [inline be/src/exec/parquet/column_reader.h:291]: `init_converter` -> `create_converter` for better semantic

Comment (DorianZheng) [inline be/src/exec/parquet/column_reader.h:291]: const ParquetField& instead of const ParquetField*

File: be/src/exec/parquet/column_reader.h [hunk 4]
```diff
@@ -27,6 +30,269 @@ struct ColumnReaderOptions {
...
+                                 const ColumnReaderOptions& opts, std::unique_ptr<ColumnConverter>* converter);
```
Comment (DorianZheng) [inline be/src/exec/parquet/column_reader.h:292]: extract `ColumnConverterOptions` instead?

File: be/src/exec/parquet/column_reader.h [hunk 5]
```diff
@@ -27,6 +30,268 @@ struct ColumnReaderOptions {
...
+};
+
+template <typename T>
+struct DecimalTypeTraits {
+    using PrimitiveType = T;
+};
+
+template <>
+struct DecimalTypeTraits<DecimalV2Value> {
+    using PrimitiveType = int128_t;
```
Comment (mofeiatwork) [inline be/src/exec/parquet/column_reader.h:49]: `PrimitiveType` already exists, such as `TYPE_INT`, `TYPE_VARCHAR`.

File: be/src/exec/parquet/column_reader.h [hunk 6]
```diff
@@ -27,6 +30,268 @@ struct ColumnReaderOptions {
...
+
+        auto& src_data = src_column->get_data();
+        auto& dst_data = dst_column->get_data();
+        auto& src_null_data = src_nullable_column->null_column()->get_data();
+        auto& dst_null_data = dst_nullable_column->null_column()->get_data();
+
+        size_t size = src_column->size();
+
+        for (size_t i = 0; i < size; i++) {
+            dst_null_data[i] = src_null_data[i];
```
Comment (mofeiatwork) [inline be/src/exec/parquet/column_reader.h:153]: Could it be replaced as a plain `memcpy`?

File: be/src/exec/parquet/column_reader.h [hunk 7]
```diff
@@ -27,6 +30,268 @@ struct ColumnReaderOptions {
...
+        for (size_t i = 0; i < size; i++) {
+            dst_data[i] = TO(src_data[i]);
+        }
+
+        dst_nullable_column->set_has_null(src_nullable_column->has_null());
+        return Status::OK();
+    }
+};
+
+template <typename SourceType, PrimitiveType dst_type>
```
Comment (mofeiatwork) [inline be/src/exec/parquet/column_reader.h:164]: wow, naming style should be consistent?

**Review by murphyatwork — DISMISSED** [review]

File: be/src/exec/parquet/column_reader.h (hunk 1 above)
Comment (miomiocat) [reply be/src/exec/parquet/column_reader.h]: done

File: be/src/exec/parquet/column_reader.h (hunk 3 above)
Comment (miomiocat) [reply be/src/exec/parquet/column_reader.h]: done

Comment (miomiocat) [reply be/src/exec/parquet/column_reader.h]: done

File: be/src/exec/parquet/column_reader.h (hunk 4 above)
Comment (miomiocat) [reply be/src/exec/parquet/column_reader.h]: use string timezone directly

File: be/src/exec/parquet/column_reader.h (hunk 5 above)
Comment (miomiocat) [reply be/src/exec/parquet/column_reader.h]: removed and revised

File: be/src/exec/parquet/column_reader.h (hunk 7 above)
Comment (miomiocat) [reply be/src/exec/parquet/column_reader.h]: revised

File: be/src/exec/parquet/column_reader.h (hunk 2 above)
Comment (miomiocat) [reply be/src/exec/parquet/column_reader.h]: done

**Review by murphyatwork — DISMISSED** [review]
LGTM

File: be/src/exec/parquet/file_reader.cpp [hunk 8]
```diff
@@ -274,7 +272,8 @@ int FileReader::_get_partition_column_idx(const std::string& col_name) const {
     return -1;
 }
 
-Status FileReader::_decode_min_max_column(const tparquet::ColumnMetaData& column_meta,
+Status FileReader::_decode_min_max_column(const ParquetField& field, const std::string& timezone,
```
Comment (DorianZheng) [inline be/src/exec/parquet/file_reader.cpp:275]: You can simplify this code by using `PlainDecoder<PhysicalTypeTraits<column_meta.type>::CppType>`

**Review by DorianZheng — DISMISSED** [review]
LGTM

File: be/src/exec/parquet/column_converter.h [hunk 9]
```diff
@@ -0,0 +1,279 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
```
Comment (imay) [inline be/src/exec/parquet/column_converter.h:2]: need include guard
#pragma once

Comment (miomiocat) [reply be/src/exec/parquet/column_converter.h:2]: done

**Comment by miomiocat** [comment]
run starrocks_be_unittest

**Comment by miomiocat** [comment]
run starrocks_be_unittest

**Review by caneGuy — APPROVED** [review]
LGTM

**Review by dirtysalt — APPROVED** [review]

**Comment by miomiocat** [comment]
@Mergifyio backport branch-2.2

**Comment by miomiocat** [comment]
@Mergifyio backport branch-2.1

**Comment by miomiocat** [comment]
@Mergifyio backport branch-2.0

**Pull request merged**